fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/lowcardinality/DecodeRewriter.java

```diff
@@ -0,0 +1,390 @@
+                    !context.stringRefToDictRefMap.containsKey(orderDesc.getColumnRef())) {
+                newOrdering.add(orderDesc);
+                continue;
+            }
+            ColumnRefOperator dictRef = context.stringRefToDictRefMap.get(orderDesc.getColumnRef());
+            newOrdering.add(new Ordering(dictRef, orderDesc.isAscending(), orderDesc.isNullsFirst()));
+        }
+
+        List<ColumnRefOperator> newPartitionByColumns = null;
+
+        if (topN.getPartitionByColumns() != null) {
+            newPartitionByColumns =
+                    topN.getPartitionByColumns().stream().map(c -> context.stringRefToDictRefMap.getOrDefault(c, c))
+                            .collect(Collectors.toList());
+        }
+
+        ScalarOperator predicate = rewritePredicate(topN.getPredicate(), info.inputStringColumns);
+        Projection projection = rewriteProjection(topN.getProjection(), info.inputStringColumns);
+        PhysicalTopNOperator newOp =
+                PhysicalTopNOperator.builder().withOperator(topN).setOrderSpec(new OrderSpec(newOrdering))
+                        .setPartitionByColumns(newPartitionByColumns).setPredicate(predicate).setProjection(projection)
+                        .build();
+        return rewriteOptExpression(optExpression, newOp, info.outputStringColumns);
+    }
+
+    @Override
+    public OptExpression visitPhysicalOlapScan(OptExpression optExpression, ColumnRefSet fragmentUseDictExprs) {
+        PhysicalOlapScanOperator scanOperator = (PhysicalOlapScanOperator) optExpression.getOp();
+        DecodeNodeInfo info = context.operatorDecodeInfo.get(scanOperator);
+
+        Map<ColumnRefOperator, Column> newRefToMetaMap = Maps.newHashMap();
+        List<Pair<Integer, ColumnDict>> dicts = Lists.newArrayList();
+        for (ColumnRefOperator ref : scanOperator.getColRefToColumnMetaMap().keySet()) {
+            Column meta = scanOperator.getColRefToColumnMetaMap().get(ref);
+
+            if (!info.inputStringColumns.contains(ref.getId())) {
+                newRefToMetaMap.put(ref, meta);
+                continue;
+            }
+
+            ColumnRefOperator dictRef = context.stringRefToDictRefMap.get(ref);
+            Column newMeta = new Column(meta);
+            newMeta.setType(dictRef.getType());
+            newRefToMetaMap.put(dictRef, newMeta);
+            dicts.add(new Pair<>(dictRef.getId(), context.stringRefToDicts.get(ref.getId())));
+        }
+
+        PhysicalOlapScanOperator.Builder builder = PhysicalOlapScanOperator.builder();
+        builder.withOperator(scanOperator);
+
+        builder.setColRefToColumnMetaMap(newRefToMetaMap);
+        builder.setGlobalDicts(dicts);
+        builder.setGlobalDictsExpr(computeDictExpr(fragmentUseDictExprs));
+
+        builder.setPredicate(rewritePredicate(scanOperator.getPredicate(), info.inputStringColumns));
+        builder.setProjection(rewriteProjection(scanOperator.getProjection(), info.inputStringColumns));
```

Comment:
How process scenes like need insert a decodeOpExpr before the projection?

Reply:
I think none, becasue expression(based on low-cardinality) compute preformance is better than based on string

Reply:
https://github.com/StarRocks/starrocks/pull/24425 how about thess case?

Reply:
I was added all old test case